stl/inc/algorithm

```diff
@@ -5035,6 +5046,238 @@ _FwdIt shift_right(_ExPo&&, _FwdIt _First, _FwdIt _Last, _Iter_diff_t<_FwdIt> _P
+            requires indirectly_movable<_It, _Out>
+        _NODISCARD static constexpr _Out _Move_n_helper(_It _First, iter_difference_t<_It> _Count, _Out _Result) {
+            // clang-format on
+            for (; _Count > 0; ++_First, (void) --_Count, ++_Result) {
+                *_Result = _RANGES iter_move(_First);
+            }
+
+            return _Result;
+        }
+
+        template <class _It, class _Se, class _SizeTy>
+        _NODISCARD static constexpr subrange<_It> _Shift_right_impl(
+            _It _First, _Se _Last, iter_difference_t<_It> _Pos_to_shift, const _SizeTy _Size) {
+            _STL_INTERNAL_STATIC_ASSERT(permutable<_It>);
+            _STL_INTERNAL_STATIC_ASSERT(sentinel_for<_Se, _It>);
+
+            if (_Pos_to_shift == 0) {
+                return {_First, _RANGES next(_First, _Last)};
+            }
+
+            constexpr bool _Is_sized = !same_as<_SizeTy, _Unsized>;
+
+            if constexpr (_Is_sized) {
+                if (_Pos_to_shift >= _Size) {
+                    if constexpr (same_as<_It, _Se>) {
+                        return {_Last, _Last};
+                    } else if constexpr (random_access_iterator<_It>) {
+                        _First += _Size;
+                        return {_First, _First};
+                    } else {
+                        _RANGES advance(_First, _Last);
+                        return {_First, _First};
+                    }
+                }
+            }
+
+            if constexpr (_Bidi_common<_It, _Se>) {
+                auto _Mid = _Last;
+                if constexpr (_Is_sized) {
+                    // We validated above that the length of the range is more than _Pos_to_shift
+                    _RANGES advance(_Mid, -_Pos_to_shift);
+                } else {
+                    if (_RANGES advance(_Mid, -_Pos_to_shift, _First) != 0) {
+                        return {_Last, _Last};
+                    }
+                }
+                return {_RANGES _Move_backward_common(_STD move(_First), _STD move(_Mid), _Last), _Last};
+            } else if constexpr (_Is_sized && random_access_iterator<_It>) {
+                auto _Final = _First + _Size;
+                auto _Mid   = _Final - _Pos_to_shift;
+                return {_RANGES _Move_backward_common(_STD move(_First), _STD move(_Mid), _Final), _Final};
+            } else {
+                auto _Buf = _First;
+                if constexpr (_Is_sized) {
+                    _RANGES advance(_Buf, _Pos_to_shift);
+                } else {
+                    if (_RANGES advance(_Buf, _Pos_to_shift, _Last) != 0) {
+                        return {_Buf, _Buf};
+                    }
+                }
+
+                if constexpr (_Is_sized) {
+                    if (_Size < 2 * _Pos_to_shift) {
+                        return {_Buf, _Move_n_helper(_First, _Size - _Pos_to_shift, _Buf)};
+                    }
+                }
+                auto _Lead = _Buf;

```

Comment:
Could we add some lines between the non chained  controlflow

Reply:
This is all one paragraph to me, with the goal of determining if `next(_Buf, _Pos_to_shift)` is valid and defining `_Lead` to that value if so, so it makes sense to me this way. I'll add a newline before the definition of `_Lead`.

Reply:
Sounds good to me. The scenario where our conventions insist on a newline is between `if`/`else if`/`else` chains: `if (is_cute) { } if (is_fluffy) { } if (is_kitten) { }`. Newlines in other situations are basically "author's preference". Personally, I like to have a newline between a condition and a variable, which is what Casey's adding here: `if (cond) { } Type var;`, since this is like "taking a breath" before starting something new. Whether there should be a newline between a variable and a following condition, as in `Type var; if (another_cond) { }` seems context-dependent to me (having a newline is a bit easier to read; omitting it signals that the variable is strongly associated with the following condition).

Thanks for always thinking about code clarity! :smile_cat: